README.md

```diff
@@ -2,4 +2,61 @@
 
 A multi-threaded implementation of a PromQL Query Engine based on the [Volcano/Iterator model](https://paperhub.s3.amazonaws.com/dace52a42c07f7f8348b08dc2b186061.pdf).
 
-The project is currently under development.
+The project is currently under active development.
+
+## Design
+
+At the beggining of a PromQL query execution, the query engine computes a physical plan consisting of multiple independent operators, each responsible for calculating one part of the query expression.
+
+Operators are assembled in a tree-like structure with every operator calling `Next()` on its dependants until there is no more data to be returned. The result of the `Next()` function is a _column vector_  (also called a _step vector_) with elements in the vector representing samples with the same timestamp from different time series.
+
+<p align="center">
+  <img src="./assets/design.png" />
+</p>
+
+This model allows for samples from individual time series to flow one execution step at a time from the left-most operators to the one at the very right. Since most PromQL expressions are aggregations, samples are reduced in number as they are pulled by the operators on the right. Because of this, samples from original timeseries can be decoded and kept in memory in batches instead of being fully expanded.
+
+In addition to operators that have a one-to-one mapping with PromQL constructs, the Volcano model also describes so-called Exchange operators which can be used for flow control and optimizations, such as concurrency or batched selects. An example of an _Exchange_ operator is described in the [Intra-operator parallelism](#intra-operator-parallelism) section.
+
+## Inter-operator parallelism
+
+Since operators are independent and rely on a common interface for pulling data, they can be run in parallel to each other. As soon as one operator has processed data from an evaluation step, it can pass the result onward so that its upstream can immediatelly start working on it.
```

Comment:
```suggestion
Since operators are independent and rely on a common interface for pulling data, they can be run in parallel to each other. As soon as one operator has processed data from an evaluation step, it can pass the result onward so that its upstream can immediately start working on it.
```